docs/cip/CIP-8_Matrix_Packages.md

```diff
@@ -0,0 +1,209 @@
+
+We want to foster an ecosystem of "knowledge packages," such that users can stitch together RAG programs by combining these packages with their own data. We will call such packages, and their format, "matrix." Drop a matrix package into your RAG app - now it knows kung fu.
+
+Many steps are necessary to reach this overarching goal, but the history of traditional software packaging clearly indicates that simple tools for both producing and consuming packages are prerequisite to healthy ecosystems. The scope of this CIP is an initial step towards that end: define a knowledge package format, and make the information held in Chroma portable into that format. The completeness criteria for this proposal is to define a full round trip, chroma -> matrix -> chroma.
+
+Ideally, the matrix package format could eventually be shared by the many tools in the emergent LLM/AI stack. This proposal seeks to balance that Chroma-agnostic goal with the reality that working, useful software is usually prerequisite to bootstrapping an ecosystem.
+## Impact
+
+This is a novel user path for Chroma, and as such does not affect any existing code or user behaviors. It may eventually affect workflows for developing and deploying Chroma-based apps, possibly even those that have no external dependencies.
+
+The main impact is on maintenance burden and (possible) compatibility guarantees. By creating what can be thought of as a portable serialization format for most of what is ordinarily stored in Chroma, it is likely that some properties which previously could have been considered implementation details, now will be assumed by users to be part of Chroma's public API surface.
+
+TODO examples
+
+## Proposed Change
+
+This is a proposal in two parts. First, we describe a package format. Then, we outline new chroma client capabilities and corresponding CLI subcommands that can round trip to and from this package format.
+
+Note that this CIP explicitly omits any actual dependency management behavior. While certain design choices here are motivated by preparing for that problem, directly addressing it is a problem for a future CIP.
+### Matrix Package Format
+A matrix package is a filesystem format consisting of two files, co-occurring within the same directory. Each file serves a particular, well-defined role:
+
+- `matrix.package.yaml`: defines the package name, version, and other common metadata, and certain metadata about the dataset in the package. Future iterations will add dependency declarations - the same kind of information that appears in `package.json` or `requirements.txt`. **Required**.
+- `matrix.data.parquet`: contains the package's dataset. Each row closely mirrors Chroma db content, containing original document source, computed embeddings, and metadata. **Required**, but will likely become optional in the future (more below)
+
+TODO - how do we serialize a computed index? it doesn't have the same cardinality as the data, does it? so that'd mean a separate file
```

Comment:
I think this is key and tricky. I think its fairly important that any notion of a package has the index shipped alongside it for fast ingest

Reply:
We also have to consider the metadata index and also consider how this would interop with #1125

Reply:
@HammadB, maybe we don't need to overoptimise this for now. My experience shows that usually, large ingestions (especially into RDBMS), people favour disabling indexing until after the operation is done. wdyt?

Reply:
I do think it would be fine to ignore the index for this initial proposal - scope of this CIP is already pretty staggeringly large. I'll get rid of that TODO.

But it's also worth situating in a larger context at least here in the comments.

For any given dataset, there's a combinatorial explosion of possible downstream computed values. Embeddings depend on/are nominally typed by the model used, the EF, and IIUC also e.g. the hnsw params. Indexes, in turn, depend on the computed embeddings, the chosen indexing algorithm, etc.

All of these are potentially expensive to calculate - in time, CPU cycles, and in $$ if you're paying by the byte for the first step in the process. And it's not clear it will ever be feasible for package authors to make [context-independent choices about what to include in their package](https://github.com/chroma-core/chroma/pull/1179/files#diff-9a9f395ccd84b350cbc87c641f80dbda9399367ade4341c88530bb06cfc973e0R205). This leads me to the conclusion that it will be normal for consumers of a package to have to (re)compute at least some of the values published with the package under different parameterizations.

If expensive [re]computation is going to be the norm, it may be worth designing around. Indexes become one instance in a pattern of downstream computed values which may or may not be published with a package.

I think this way of looking at it is in line with [this note](https://github.com/chroma-core/chroma/pull/1110/files#diff-66f22f80078f4d1dd1e86f8d14763937cd567e27f2664cc97d5110d2a8bcfd6eR295-R296) from CIP 6:

> Collections are just a description input pipelines, query pipelines and output pipelines. Today, the pipeline runs an EF and stores it in two segments.
Everything that runs before writing to log is the input pipeline, everything that runs after vector search is the output pipeline, and the query pipeline merely transforms the query.

Reply:
Wrt to hnsw indices, I think there is also the consideration of portability across platforms.